### ISSUE TITLE
Clarify DWARF Version operand of DebugCompilationUnit

### ISSUE THREAD

**Issue by MrSidims** [KhronosGroup/SPIRV-Registry#280]
On windows default debug info format is CodeView and we would like to have an ability to express CodeView version in SPIR-V. For [OpenCL.DebugInfo.100](https://registry.khronos.org/SPIR-V/specs/unified1/OpenCL.DebugInfo.100.html) there are not much alternatives but to set zero as `DWARF Version` literal operand of `DebugCompilationUnit`. Starting from [NonSemantic.Shader.DebugInfo.100](https://github.com/KhronosGroup/SPIRV-Registry/blob/main/nonsemantic/NonSemantic.Shader.DebugInfo.100.asciidoc) there are potential alternatives.

Currently our SPIR-V producer ([LLVM IR to SPIR-V translator](https://github.com/KhronosGroup/SPIRV-LLVM-Translator)) generates zero constant as `DWARF Version` operand of `DebugCompilationUnit` for CodeView and it becomes hard for the SPIR-V consumer to decide on the debug info version to set.

The questions/proposals I have:
1. Can we modify `DebugCompilationUnit` within NonSemantic.Shader.DebugInfo.100 to allow `DWARF Version` to also be `OpString`? Like this we could encode not only DWARF, but any other potential debug info formats.
2. If the answer to the above proposal is "no", then what should be the value of DWARF Version for non-DWARF debug info formats? On the one hand the specification says, that `DWARF Version` must be a 32-bit integer constant (and so in the translator we generate OpConstant with zero value), on the other hand [DebugInfoNone](https://github.com/KhronosGroup/SPIRV-Registry/blob/main/nonsemantic/NonSemantic.Shader.DebugInfo.100.asciidoc#DebugInfoNone) "Other instructions can refer to this one in case the debugging information is unknown, not available, or not applicable.", which seems to be a perfect fit in this case.

**Comment by MrSidims** [comment]
@baldurk hi, may I ask you to express your opinion on that?

**Comment by baldurk** [comment]
This was ported directly from the OpenCL spec, as I prioritised making as few breaking changes as possible since the original goal was to make some code that generated debug info usable on vulkan. I don't really understand the purpose of this field or what SPIR-V tools could feasibly use it for, but we definitely can't make backwards-compatible breaking changes.

All the spec says is that this field is the DWARF version that "this specification" is compatible with, meaning the debuginfo itself, but it doesn't define that compatibility and there aren't any external dependencies on DWARF as the spec is self-contained so I'm not sure what that could mean. Unless there's a meaningful interpretationn it seems like every emitter could set this field to '5' to refer to DWARF version 5 and leave it at that, I don't see a more useful value to put there.

I don't understand what you mean by "express CodeView version in SPIR-V" or that you are forced to emit 0 for this field. Whether you separately generate codeview info elsewhere is unrelated to the SPIR-V DebugInfo as they are two different specs. If you wanted to add some codeview information to SPIR-V it would need a separate spec and extension.

**Comment by VyacheslavLevytskyy** [comment]
It looks like https://github.com/KhronosGroup/SPIRV-Registry/blob/main/nonsemantic/NonSemantic.Shader.DebugInfo.100.asciidoc#missing-debugging-information has so global impact on the specification that it deserves even more attention and clarification, probably within the Introduction. Otherwise there will always be doubts if DebugInfoNone is indeed a valid substitution for each <id> in every instruction.

**Comment by baldurk** [comment]
`DebugInfoNone` is not globally applicable it's used in specific cases where there may be an ID with no valid value, like a pointer that can point to NULL. These cases are called out specifically - e.g. `DebugTypeEnum` if there's no underlying type, or `DebugGlobalVariable` if there's no variable.

**Comment by VyacheslavLevytskyy** [comment]
Thank you @baldurk . Am I correct saying:

"if there is no explicit statement in the description of the Instruction that Id may be substituted by DebugInfoNone, it's not allowed to use DebugInfoNone instead of this Id" ?

**Comment by baldurk** [comment]
Yes that's how I understand it, unless there's some missed case where an id is referenced that might need to be allowed as None then it would only be where specified.

**Comment by MrSidims** [comment]
> Yes that's how I understand it, unless there's some missed case where an id is referenced that might need to be allowed as None then it would only be where specified.

That's actually quite surprising for me. If we look at DWARF 5 specification, lets say on the following paragraph:

> **2.14 Declaration Coordinates**
> Any debugging information entry representing the declaration of an object,
> module, subprogram or type **may have** DW_AT_decl_file, DW_AT_decl_line and
> DW_AT_decl_column attributes, each of whose value is an unsigned integer
> 16 constant.

Now going to, lets say, a subprogram declaration in SPIR-V debug info specification, **DebugFunctionDeclaration**:

> Column is the <id> of a 32-bit integer OpConstant denoting the column number at which the first character of the function declaration appears.

So whilst column information would be an optional information in DWARF - it becomes mandatory in SPIR-V.

**Comment by baldurk** [comment]
I'm not familiar with DWARF but I don't think that specification is relevant here as we're talking about the SPIR-V NonSemantic.Shader.DebugInfo.100.

The convention I've seen for column information if it's not available is to set both start and end to 0, though you're right that this should be stated explicitly.

The change from literals was not so that IDs could be substituted with DebugInfoNone, indeed literals would be much simpler for most cases they're used for. It was necessary due to the original requirement from the SPIR-V working group that non-semantic extended instruction sets have IDs for all parameters so that tools could theoretically process them without understanding them.

**Comment by MrSidims** [comment]
> I don't think that specification is relevant here as we're talking about the SPIR-V NonSemantic.Shader.DebugInfo.100.

Agree. And in a flow SPIR-V -> (dwarf producer) -> DWARF it's fine to put stricter rules for SPIR-V.

> The convention I've seen for column information if it's not available is to set both start and end to 0, though you're right that this should be stated explicitly.

I'm a bit afraid that we would need to specify quite a lot of defaults then, when debug information is not available. On the other hand it's better, then to blindly rely on DebugInfoNone leaving DWARF producer to decide, what attribute to pick (if applicable).

**Comment by baldurk** [comment]
Yes I agree blanket allowing DebugInfoNone all over the place would make the situation significantly harder for consumers to handle normal cases. I don't think producing DWARF information from SPIR-V's NonSemantic debug info was ever a goal, that seems like it would have to be 'best effort' if the information is there.

If there are other situations where information could reasonably be optional that seems like a case-by-case basis for spec clarification but I would hope they are rare - aside from fields like in the original issue where it's inherited from the OpenCL debuginfo and doesn't make sense.

Honestly I would personally say that column information shouldn't be optional either, at worst it should just refer to the whole length of the source line if no sub-expression column information is available, but it's a common one for frontends not to generate so there's the convention of leaving 0.

**Comment by MrSidims** [comment]
> but it's a common one for frontends not to generate so there's the convention of leaving 0.

Sure, setting column number to zero is fine (that's actually what we already do in LLVM IR to SPIR-V translator).

I'll scrub, what we have done for https://github.com/KhronosGroup/SPIRV-LLVM-Translator - there are not many places, where DebugInfoNone is being generated to see, where it can be defaulted. If there are objectively impossible to substitute info with some default value, I'll open a PR to the spec and ask you to review :)

And thank you for the answers!

**Comment by bwlodarcz** [comment]
> `DebugInfoNone` is not globally applicable it's used in specific cases where there may be an ID with no valid value, like a pointer that can point to NULL. These cases are called out specifically - e.g. `DebugTypeEnum` if there's no underlying type, or `DebugGlobalVariable` if there's no variable.

@baldurk If this is true then how we are able e.g. implement `void *` type for `DebugTypePointer`. It doesn't have the information specified that I can substitute it as `DebugInfoNone` in spec and currently I don't see any way to avoid it.

It's a strange approach to define places in specification which can be substituted by Null because it's certain that real world will find at least some unforeseen by spec definition places which are perfect unavoidable match - and when that's case the implementations will diverge due to the workarounds which isn't desirable from standard PoV.

**Comment by baldurk** [comment]
Seems to me like a sensible solution would be to allow referring to `OpTypeVoid` for the base type in `DebugTypePointer` in this case, the same way that `DebugTypeFunction` allows doing that if the return type is void without giving a way to represent void as a `DebugType*`.

Or you could use define void via a `DebugTypeBasic` which is `Unspecified` encoding with the `FlagUnknownPhysicalLayout` flag.

To be clear I'd say this is another reasonable case where `DebugInfoNone` could be called out in the spec, but I still think that allowing any emitter to replace any debug info with `DebugInfoNone` at any time with no direction on what that means by calling it out explicitly is a poor idea because it means consumers can't count on any information always being present and will potentially have to make up what it means if the information is missing.

**Comment by bwlodarcz** [comment]
@baldurk Thank you for the answer and for possible solutions. I'll implement one of them.

However referring to general discussion about DebugInfoNone the answer on this question is what actually proves of my point.
If there exist more then one workaround on the problem that means that given enough time the both workarounds will be implemented.
Now, every client of the SPIR-V must have control flow for both workarounds instead one for the default `DebugInfoNone` as custom implementation defined one. Multiply divergence by number of possible workarounds for every such scenario and the complexity shots over the roof. 
I think that it's better for ecosystem and standard to define 'Implantation defined when no data' instruction than forcing workarounds for non covered by standard lack of data scenario.

**Comment by baldurk** [comment]
I don't understand your argument. Yes there's a bug/ambiguity in the spec on how to represent something - you commented about it here rather than opening a new issue but it should still certainly be fixed. I don't see where workarounds come into it at all.

Also to repeat again, I don't have a problem with using `DebugInfoNone` as long as it's called out in the spec and the meaning of it is detailed in context so that consumers know what to do with it. Allowing it anywhere for any debug id without specifying what it means consumers need to add error handling for every single id and then either can't use the debug info, or will have to make up their own interpretations of what it means.

**Comment by bwlodarcz** [comment]
@baldurk Ok my bad. Give me one more try ;)
 
Currently e.g. SPIR-V Translator based a lot about the fact the `DebugInfoNone` is applicable in places where there aren't defined data. After the discussion above with @MrSidims and @VyacheslavLevytskyy it seems clear that `DebugInfoNone` should be used only in a places defined by instruction definitions in spec which renders vast amounts of Translator implementation incorrect.
I'm currently implementing DI in SPIR-V Backend in llvm - it's a second time when DebugInfoNone is actually (IMO) natural clean and easy solution but because currently isn't described in instruction description is rendered as non compliant.
I'm expecting, maybe incorrectly, that during the implementation I will periodically hit similar scenarios.

I'm not considering `void *` case as a oversight and bug in spec because it's natural to not see every possibility. Spec author 
can't see every possible use case and conditions. 
I'm considering a rule which REQUIRE foreseeing such cases as being not thought thorough enough. We will endlessly in need to add exceptions and edge cases to defined rules or creating workarounds in implementations just to avoid using Null. To define what's I'm considering workaround here - avoiding simple solution and creating synthetic additional instructions or control flows just to fulfill spec to the letter. Typically workaround characterizes itself as more complex and less performant then obvious solution. 
There are two possible outcomes here:
a) When someone hit a explained case will create a workaround in it's implementation
b) Will add an exception to the standard like we discussing now

b) is slower to do than a) (in time allocation, it's easier to implement and merge then communicating about standard) so you have guarantee that the implementations will diverge.
Even if we consider that a < b in time allocation then giving time you are ending with spec where every possible instruction have Null edge case. Considering that scenario you can just go ahead and make it legal now.

Now from purely consumer point of view. Let's say that I need to consume SPIR-V from every available on my platform producer and have two workarounds presented above.
Producer A uses OpTypeVoid solution.
Producer B uses DebugTypeBasic solution.
As a consumer you must handle and validate both. With `DebugInfoNone` there I, as a consumer, interpret it as 'lack of data' and 'implementation defined' which means that I'm free to handle this as I want. Both producers sending same input which is simpler to handle and understand. 

I hope that I explained my point well enough this time. 
If it's more desirable to start separate issue for the topic I can create it.

**Comment by Tobski** [comment]
@bwlodarcz this spec was really just an adjusted version of the OpenCL debug semantics, which was done pragmatically to meet the fact that those semantics were already being used in tooling but couldn't be consumed by Vulkan implementations. That was, more or less, the intended use case.

It sounds like it doesn't fit your use case as well as you'd hoped, it just happens to be the only thing that exists. If that's the case, it might be time to consider defining a new non-semantic instruction set that is properly designed for Vulkan SPIR-V and the use cases that we have.

I would strongly recommend raising a new issue for this if you want a change to the existing spec; if there's enough issues that you might be happier with a new instruction set, then I'd suggest you raise a new issue with a list of your requirements/nice to haves/issues with the current instruction set.
